redshift_sqlalchemy/dialect.py

```diff
@@ -320,67 +323,157 @@ def visit_unload_from_select(element, compiler, **kw):
+                             ['CSV', 'JSON', 'AVRO'])
+
+        if compression is not None and compression not in ['GZIP', 'LZOP']:
+            raise ValueError('"compression" parameter must be one of %s' %
+                             ['GZIP', 'LZOP'])
+
+        self.table = table
         self.data_location = data_location
-        self.access_key = access_key
-        self.secret_key = secret_key
-        self.session_token = session_token
-        self.options = options
+        self.credentials = credentials
+        self.delimiter = delimiter
+        self.ignore_header = ignore_header
+        self.null = null
+        self.manifest = manifest
+        self.empty_as_null = empty_as_null
+        self.blanks_as_null = blanks_as_null
+        self.format = format
+        self.compression = compression or ''
+
+
+class _UnquotedParameter(sa.types.TypeDecorator):
+    """Custom type for not quoting a BindParameter
+    """
+    impl = sa.types.UserDefinedType
+
+    def process_bind_param(self, value, dialect):
+        return value
 
 
 @compiles(CopyCommand)
 def visit_copy_command(element, compiler, **kw):
     ''' Returns the actual sql query for the CopyCommand class
     '''
-    return """
-           COPY %(schema_name)s.%(table_name)s FROM '%(data_location)s'
-           CREDENTIALS 'aws_access_key_id=%(access_key)s;aws_secret_access_key=%(secret_key)s%(session_token)s'
-           CSV
-           TRUNCATECOLUMNS
-           DELIMITER '%(delimiter)s'
-           IGNOREHEADER %(ignore_header)s
-           %(null)s
-           %(manifest)s
-           %(empty_as_null)s
-           %(blanks_as_null)s;
-           """ % \
-           {'schema_name': element.schema_name,
-            'table_name': element.table_name,
-            'data_location': element.data_location,
-            'access_key': element.access_key,
-            'secret_key': element.secret_key,
-            'session_token': ';token=%s' % element.session_token if element.session_token else '',
-            'null': ("NULL '%s'" % element.options.get('null')) if element.options.get('null') else '',
-            'delimiter': element.options.get('delimiter', ','),
-            'ignore_header': element.options.get('ignore_header', 0),
-            'manifest': 'MANIFEST' if bool(element.options.get('manifest', False)) else '',
-            'empty_as_null': 'EMPTYASNULL' if bool(element.options.get('empty_as_null', True)) else '',
-            'blanks_as_null': 'BLANKSASNULL' if bool(element.options.get('blanks_as_null', True)) else ''}
+    qs = """COPY :schema.:table FROM :data_location
+    CREDENTIALS :credentials
+    :format
+    TRUNCATECOLUMNS
+    DELIMITER :delimiter
+    IGNOREHEADER :ignore_header
+    NULL :null
+    :manifest
+    :compression
+    :empty_as_null
+    :blanks_as_null
+    """
+
+    return compiler.process(
+        sa.text(qs).bindparams(
+            sa.bindparam('schema',
+                         value=element.table.schema or 'public',
```

Comment:
you shouldn't assume 'public' here, it needs to run against the currently selected schema. It should just be left out if not specified